tap-agent/src/tap/sender_account.rs

```diff
@@ -0,0 +1,921 @@
+        let mut allocations_to_move = Vec::new();
+        for allocation_id in allocations_active.keys() {
+            if !target_allocations.contains(allocation_id) {
+                allocations_to_move.push(*allocation_id);
+            }
+        }
+        for allocation_id in &allocations_to_move {
+            allocations_ineligible.insert(
+                *allocation_id,
+                allocations_active.remove(allocation_id).unwrap(),
+            );
+        }
+
+        // If we moved any allocations to the ineligible map, notify the RAV requester finalize
+        // task.
+        if !allocations_to_move.is_empty() {
+            self.rav_requester_finalize_notify.notify_waiters();
+        }
+
+        // Add new allocations.
+        for allocation_id in target_allocations {
+            if !allocations_active.contains_key(&allocation_id)
+                && !allocations_ineligible.contains_key(&allocation_id)
+            {
+                allocations_active.insert(
+                    allocation_id,
+                    Arc::new(
+                        SenderAllocation::new(
+                            self.inner.config,
+                            self.inner.pgpool.clone(),
+                            allocation_id,
+                            self.inner.sender,
+                            self.escrow_accounts.clone(),
+                            self.escrow_subgraph,
+                            self.escrow_adapter.clone(),
+                            self.tap_eip712_domain_separator.clone(),
+                            self.inner.sender_aggregator_endpoint.clone(),
+                        )
+                        .await,
+                    ),
+                );
+            }
+        }
+    }
+
+    pub async fn handle_new_receipt_notification(
+        &self,
+        new_receipt_notification: NewReceiptNotification,
+    ) {
+        let mut unaggregated_fees = self.inner.unaggregated_fees.lock().await;
+
+        // Else we already processed that receipt, most likely from pulling the receipts
+        // from the database.
+        if new_receipt_notification.id > unaggregated_fees.last_id {
+            if let Some(allocation) = self
+                .inner
+                .allocations_active
+                .lock()
+                .await
+                .get(&new_receipt_notification.allocation_id)
+            {
+                // Add the receipt value to the allocation's unaggregated fees value.
+                allocation.fees_add(new_receipt_notification.value).await;
+                // Add the receipt value to the sender's unaggregated fees value.
+                Self::fees_add(
+                    self.inner.clone(),
+                    &mut unaggregated_fees,
+                    new_receipt_notification.value,
+                );
+
+                unaggregated_fees.last_id = new_receipt_notification.id;
```

Comment:
fees added + last_id seems to me a whole update of the struct. Why interior mutability here?

Reply:
I'm not quite sure how I should change this, but I did do some simplification here: [8c6010d](https://github.com/graphprotocol/indexer-rs/pull/113/commits/8c6010d3fa041d18113f8c226060439fd779507e)
Could you elaborate if that's not what you had in mind?